File: python-midonetclient/src/midonetclient/federation/ovsdb_binding.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2015 Midokura SARL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from midonetclient import resource_base
from midonet_binding import MidonetBinding
from midonetclient.federation import vendor_media_type

LOG = logging.getLogger(__name__)


class OvsdbBinding(resource_base.ResourceBase):
    media_type = vendor_media_type.OVSDB_BINDING_JSON

    def __init__(self, uri, dto, auth):
        super(OvsdbBinding, self).__init__(uri, dto, auth)

    def get_id(self):
        return self.dto['id']

    def get_vtep(self):
        return self.dto['vtepId']

    def vtep(self, vtepId):
        self.dto['vtepId'] = vtepId
        return self

    def get_segment(self):
        return self.dto['segmentId']

    def segment(self, segmentId):
        self.dto['segmentId'] = segmentId
        return self

    def get_port_name(self):
        return self.dto['portName']

    def port_name(self, portName):
        self.dto['portName'] = portName
        return self

    def get_vlan(self):
        return self.dto['vlanId']

    def vlan(self, vlanId):
        self.dto['vlanId'] = vlanId
        return self

    def get_router_mac(self):
        return self.dto['routerMac']

    def router_mac(self, routerMac):
        self.dto['routerMac'] = routerMac
        return self

    def get_router_cidr(self):
        return self.dto['routerCidr']

    def router_cidr(self, routerCidr):
        self.dto['routerCidr'] = routerCidr
        return self

    def get_local_subnets(self):
        return self.dto['localSubnets']

    def local_subnets(self, localSubnets):
        self.dto['localSubnets'] = localSubnets
        return self

File: cluster/midonet-cluster/src/main/java/org/midonet/cluster/rest_api/models/federation/OvsdbBinding.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.rest_api.models.federation;

import java.net.URI;
import java.util.List;
import java.util.UUID;

import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;

import com.fasterxml.jackson.annotation.JsonIgnore;

import org.midonet.cluster.data.ZoomClass;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Federation;
import org.midonet.cluster.rest_api.models.UriResource;
import org.midonet.cluster.rest_api.validation.MessageProperty;
import org.midonet.cluster.util.IPAddressUtil;
import org.midonet.cluster.util.IPSubnetUtil;
import org.midonet.cluster.util.UUIDUtil.Converter;
import org.midonet.packets.IPv4Subnet;
import org.midonet.packets.MAC;

@ZoomClass(clazz = Federation.OvsdbRouterBinding.class)
public class OvsdbBinding extends UriResource {

    @ZoomField(name = "id", converter = Converter.class)
    public UUID id;

    @NotNull
    @ZoomField(name = "vtep_id", converter = Converter.class)
    public UUID vtepId;

    @NotNull
    @ZoomField(name = "segment_id", converter = Converter.class)
    public UUID segmentId;

    @NotNull
    @ZoomField(name = "port_name")
    public String portName;

    @Min(0)
    @Max(4095)
    @ZoomField(name = "vlan_id")
    public short vlanId;

    @Pattern(regexp = MAC.regex, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "router_mac")
    public String routerMac;

    @NotNull
    @Pattern(regexp = IPv4Subnet.IPV4_CIDR_PATTERN, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "router_cidr", converter = IPSubnetUtil.Converter.class)
    public String routerCidr;

    @NotNull
    //@Pattern(regexp = IPv4Subnet.IPV4_CIDR_PATTERN, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "local_subnets", converter = IPSubnetUtil.Converter.class)
    public List<String> localSubnets;

    @Override
    public URI getUri() {
        return absoluteUri(Application.OVSDB_BINDINGS, id);
    }

    @JsonIgnore
    @Override
    public void create() {
        if (null == id) {
            id = UUID.randomUUID();
        }
    }

    @JsonIgnore
    public void update(OvsdbBinding from) {
        this.id = from.id;
    }
}
